### PR TITLE
Authentication for subscriptions

### PR DESCRIPTION
**Pull request by Ralls0**
# Description

In this pull request there are some changes to integrate the authentication for subscriptions

### CODE REVIEW

File: qlkube/src/decorateSubscription.js [hunk 1]
```diff
@@ -92,29 +106,79 @@ function decorateSubscription(baseSchema, targetType, enumType) {
...
+  if (!token) throw 'Parameter token cannot be empty!';
+  if (!resource) throw 'Parameter resource cannot be empty!';
+  if (!namespace) throw 'Parameter namespace cannot be empty!';
+
+  const keyCache = `${token}_${group}_${resource}_${namespace}_${name}`;
+  const lastSub = cacheSubscriptions[keyCache];
+  const canUserWatchResourceCached =
+    lastSub &&
+    !(
+      Date.now() - lastSub > 10 * 60 * 1000 &&
```
Comment (GabriFila) [inline qlkube/src/decorateSubscription.js:125]: Define a constant `TEN_SECONDS` at the start of the file to make the code more readable. (do it for other places where you compute a certain amount of time)

File: qlkube/src/decorateSubscription.js [hunk 2]
```diff
@@ -92,29 +106,79 @@ function decorateSubscription(baseSchema, targetType, enumType) {
...
+    if (canUserWatchResource) {
+      cacheSubscriptions[keyCache] = Date.now();
+      return true;
+    }
+  }
+  return false;
+}
+
+function clearCache() {
+  const courrentTimestamp = Date.now();
```
Comment (GabriFila) [inline qlkube/src/decorateSubscription.js:150]: Just a spelling mistake

File: qlkube/src/index.js [hunk 3]
```diff
@@ -47,6 +46,11 @@ async function main() {
       path: '/subscription',
       onConnect: (connectionParams, webSocket, context) => {
         console.log('Connected!');
+        const token =
+          (connectionParams.authorization &&
+            connectionParams.authorization.split(' ')[1]) ||
+          '';
+        return { token };
```
Comment (GabriFila) [inline qlkube/src/index.js:53]: If the token is not defined I would return `false`

File: qlkube/src/index.js [hunk 4]
```diff
@@ -55,7 +59,8 @@ async function main() {
 
     context: ({ req, connection }) => {
       if (connection) {
-        return {};
+        const token = connection.context.token || '';
```
Comment (GabriFila) [inline qlkube/src/index.js:62]: I would not provide a fallback

File: qlkube/README.md [hunk 5]
```diff
@@ -40,3 +40,21 @@ yarn dev
 ## Schema
 
 The generated graphql schema is served at /schema
+
+## Playground adjust
+
+A issue could be seen in the playground when you test the subscription due to some graphql-playground modules. This bug is about the scrolling of the responses. At the moment the only way to overcome this issue is to use a browser extension that inject css into the page. A possible extension for fixing that is [Stylus](https://chrome.google.com/webstore/detail/stylus/clngdbkpkpeebahjckkjfobafhncgmne).
```
Comment (giorio94) [inline qlkube/README.md:46]:
```suggestion
An issue could be seen in the playground when you test the subscription due to some graphql-playground modules. This bug is about the scrolling of the responses. At the moment, the only way to overcome this issue is to use a browser extension that injects custom css into the page. A possible extension for fixing that is [Stylus](https://chrome.google.com/webstore/detail/stylus/clngdbkpkpeebahjckkjfobafhncgmne).
```

File: qlkube/README.md [hunk 6]
```diff
@@ -40,3 +40,21 @@ yarn dev
...
+
+After you have installed the extension and opened the option, you can write a new style just click on the respective button under the menu **Actions** and add the code below:
```
Comment (giorio94) [inline qlkube/README.md:48]:
```suggestion
After you have installed the extension and opened the option panel, you can add custom css clicking on the respective button under the menu **Actions** and adding the code below:
```

**Review by giorio94 — APPROVED** [review]
AFAU, /lgtm

**Review by GabriFila — APPROVED** [review]

**Pull request merged**